hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -169,23 +179,107 @@ object HoodieSparkSqlWriter {
+              // will rebase incoming batch onto the table's latest schema (ie, reconcile them)
+              //
+              // NOTE: Since we'll be converting incoming batch from [[sourceSchema]] into [[latestTableSchema]]
+              //       we're validating in that order (where [[sourceSchema]] is treated as a reader's schema,
+              //       and [[latestTableSchema]] is treated as a writer's schema)
+              latestTableSchema
+            } else {
+              log.error(
+                s"""
+                   |Failed to reconcile incoming batch schema with the table's one.
+                   |Incoming schema ${sourceSchema.toString(true)}
+
+                   |Table's schema ${latestTableSchema.toString(true)}
+
+                   |""".stripMargin)
+              throw new SchemaCompatibilityException("Failed to reconcile incoming schema with the table's one")
+            }
+          } else {
+            // Before validating whether schemas are compatible, we need to "canonicalize" source's schema
+            // relative to the table's one, by doing a (minor) reconciliation of the nullability constraints:
+            // for ex, if in incoming schema column A is designated as non-null, but it's designated as nullable
+            // in the table's one we want to proceed w/ such operation, simply relaxing such constraint in the
+            // source schema.
+            val canonicalizedSourceSchema = AvroSchemaEvolutionUtils.canonicalizeColumnNullability(sourceSchema, latestTableSchema)
+            // In case reconciliation is disabled, we have to validate that the source's schema
+            // is compatible w/ the table's latest schema, such that we're able to read existing table's
+            // records using [[sourceSchema]].
+            if (TableSchemaResolver.isSchemaCompatible(latestTableSchema, canonicalizedSourceSchema)) {
+              canonicalizedSourceSchema
+            } else {
+              log.error(
+                s"""
+                   |Incoming batch schema is not compatible with the table's one.
+                   |Incoming schema ${canonicalizedSourceSchema.toString(true)}
+                   |Table's schema ${latestTableSchema.toString(true)}
+                   |""".stripMargin)
+              throw new SchemaCompatibilityException("Incoming batch schema is not compatible with the table's one")
+            }
+          }
+      }
+
+      validateSchemaForHoodieIsDeleted(writerSchema)
+
+      // NOTE: PLEASE READ CAREFULLY BEFORE CHANGING THIS
+      //       We have to register w/ Kryo all of the Avro schemas that might potentially be used to decode
+      //       records into Avro format. Otherwise, Kryo wouldn't be able to apply an optimization allowing
+      //       it to avoid the need to ser/de the whole schema along _every_ Avro record
+      val targetAvroSchemas = sourceSchema +: writerSchema +: latestTableSchemaOpt.toSeq
```

Comment:
It looks like this schema reconcilation logic is not specific to Spark.  Could we extract it and allow Flink/Java engine to leverage it as well?

Reply:
Good call! Am going to extract it as a standalone method (we can take the unification as a follow-up)